src/domain/ln-invoice.ts

```diff
@@ -0,0 +1,49 @@
+import lightningPayReq from "bolt11"
+import { LnInvoiceDecodeError } from "./errors"
+
+const safeDecode = (
+  bolt11EncodedInvoice,
+): lightningPayReq.PaymentRequestObject | LnInvoiceDecodeError => {
+  try {
+    return lightningPayReq.decode(bolt11EncodedInvoice)
+  } catch (err) {
+    return new LnInvoiceDecodeError(err)
+  }
+}
+
+export const decodeInvoice = (
+  bolt11EncodedInvoice: string,
+): LnInvoice | LnInvoiceDecodeError => {
+  const decodedInvoice = safeDecode(bolt11EncodedInvoice)
+
+  if (decodedInvoice instanceof Error) return decodedInvoice
+
+  let paymentHash: PaymentHash | null = null,
+    paymentSecret: PaymentSecret | null = null
+
+  decodedInvoice.tags.forEach((tag) => {
+    if (tag.tagName === "payment_hash") {
+      if (typeof tag.data === "string") {
+        paymentHash = tag.data as PaymentHash
+      } else {
+        return new LnInvoiceDecodeError("Irregular payment_hash")
+      }
+    }
+    if (tag.tagName === "payment_secret") {
+      if (typeof tag.data === "string") {
+        paymentSecret = tag.data as PaymentSecret
+      } else {
+        return new LnInvoiceDecodeError("Irregular payment_secret")
+      }
+    }
+    if (!paymentHash || !paymentSecret) {
```

Comment:
This looks like it would always fail as there are more tags than those 2. For those to keep looping one of these tags has to be first. Does the spec determine and order here?

Reply:
Good catch (pun intended). That's a duplicate check that's in the wrong place. Removed.